### PR TITLE
Allow fakers to build on previously populated fields.

### PR DESCRIPTION
**Pull request by prescod**
Allow fakers to build on previously populated fields, such as emails based on firstnames. 

Currently: Emails and Usernames are based on FirstNames and LastNames.

Also adds a bit of test coverage.

### CODE REVIEW

**Review by boakley — CHANGES_REQUESTED** [review]
We have a video call where we went over a few small changes. Let me know when you've done them and I'll make a final pass. Otherwise, this looks good to me.

File: snowfakery/fakedata/fake_data_generator.py [hunk 1]
```diff
@@ -1,22 +1,46 @@
 from difflib import get_close_matches
 import typing as T
+import random
+from snowfakery.plugins import PluginContext
+
+
+email_templates = [  # .format language doesn't allow slicing. :(
+    f"{first_name}{first_name_separator}{{lastname}}{year}@{{domain}}"
+    for first_name in ("{firstname}", "{firstname[0]}", "{firstname[0]}{firstname[1]}")
```
[the remark this answers is not in the document]
Comment (prescod) [reply snowfakery/fakedata/fake_data_generator.py:9]: I'll pad firstname to be sure.

File: snowfakery/fakedata/fake_data_generator.py [hunk 2]
```diff
@@ -1,22 +1,46 @@
...
 
 from faker import Faker, Generator
 
 
 class FakeNames(T.NamedTuple):
     f: Faker
+    faker_context: PluginContext = None
 
-    def user_name(self):
+    def user_name(self, matching: bool = True):
```
[the remark this answers is not in the document]
Comment (prescod) [reply snowfakery/fakedata/fake_data_generator.py:21]: i'll leave it undocumented until someone asks for the feature.

File: snowfakery/fakedata/fake_data_generator.py [hunk 3]
```diff
@@ -1,22 +1,46 @@
...
         """Email address using one of the "example" domains"""
+        already_created = self._already_have(("firstname", "lastname"), matching)
+        if all(already_created):
+            template = random.choice(email_templates)
+
+            return template.format(
+                firstname=already_created[0],
+                lastname=already_created[1],
+                domain=self.f.safe_domain_name(),
+                year=str(random.randint(1955, 2020)),
```
[the remark this answers is not in the document]
Comment (prescod) [reply snowfakery/fakedata/fake_data_generator.py:42]: i'll do that.

**Review by boakley — APPROVED** [review]
After having gone over the code with Paul, and seeing these last few changes, I'm satisfied.

**Review by davisagli — COMMENTED** [review]
Are any updates to documentation needed?

**Comment by prescod** [comment]
@davisagli : I think I'd just consider this "making fake data more realistic" and not needing of docs unless it causes some confusion or problem for someone.

I'll also wait on documenting the new plugin local variables feature until someone asks for a "feature like local variables for plugins". Premature documentation locks me into a particular interface.

**Pull request merged**